rethinkdb/trio_net/net_trio.py

```diff
@@ -0,0 +1,507 @@
+                if cursor is not None:
+                    cursor._extend(buf)
+                elif token in self._user_queries:
+                    # Do not pop the query from the dict until later, so
+                    # we don't lose track of it in case of an exception
+                    query, future = self._user_queries[token]
+                    res = Response(token, buf,
+                                   self._parent._get_json_decoder(query))
+                    if res.type == p_response.SUCCESS_ATOM:
+                        future.set_result(maybe_profile(res.data[0], res))
+                    elif res.type in (p_response.SUCCESS_SEQUENCE,
+                                      p_response.SUCCESS_PARTIAL):
+                        cursor = TrioCursor(self, query, res, nursery=self._nursery)
+                        future.set_result(maybe_profile(cursor, res))
+                    elif res.type == p_response.WAIT_COMPLETE:
+                        future.set_result(None)
+                    elif res.type == p_response.SERVER_INFO:
+                        future.set_result(res.data[0])
+                    else:
+                        future.set_exception(res.make_error(query))
+                    del self._user_queries[token]
+                elif not self._closing:
+                    raise ReqlDriverError("Unexpected response received.")
+        except Exception as ex:
+            if not self._closing:
+                await self.close(exception=ex)
+        finally:
+            self._reader_ended_event.set()
+
+
+class Connection(ConnectionBase):
+    def __init__(self, *args, **kwargs):
+        super(Connection, self).__init__(ConnectionInstance, *args, **kwargs)
+        try:
+            self.port = int(self.port)
+        except ValueError:
+            raise ReqlDriverError("Could not convert port %s to an integer." % self.port)
+
+    async def _stop(self, cursor):
+        self.check_open()
+        query = Query(p_query.STOP, cursor.query.token, None, None)
+        return await self._instance.run_query(query, True)
+
+    async def reconnect(self, noreply_wait=True, timeout=None):
+        await self.close(noreply_wait)
+        self._instance = self._conn_type(self, **self._child_kwargs)
+        return await self._instance.connect(timeout)
+
+    async def close(self, noreply_wait=True):
+        if self._instance is None:
+            return None
+        return await ConnectionBase.close(self, noreply_wait=noreply_wait)
+
+
+class AsyncTrioConnectionContextManager:
+    def __init__(self, *args, **kwargs):
+        self._args = args
+        self._kwargs = kwargs
+        self._conn = None
+
+    @classmethod
+    def open(cls, *args, **kwargs):
+        return cls(*args, **kwargs)
+
+    async def __aenter__(self):
+        self._conn = await connect(*self._args, **self._kwargs)
+        return self._conn
+
+    async def __aexit__(self, self, exc_type, exc, traceback):
```

Comment:
Double self here. Otherwise looks good to merge.

Reply:
Ahh yes, I missed a few other issues, too. Just pushed a commit to fix those.

*Please keep this PR open for a bit longer!* I am still integrating it into my own app, and I'm not in a rush to have this merged.

Reply:
@mehaase What do you mean by longer? Can we merge it when all the checks are passes?